tsig.go

```diff
@@ -119,25 +131,33 @@ func TsigGenerate(m *Msg, secret, requestMAC string, timersOnly bool) ([]byte, s
 	}
 
 	t := new(TSIG)
-	var h hash.Hash
-	switch CanonicalName(rr.Algorithm) {
-	case HmacSHA1:
-		h = hmac.New(sha1.New, rawsecret)
-	case HmacSHA224:
-		h = hmac.New(sha256.New224, rawsecret)
-	case HmacSHA256:
-		h = hmac.New(sha256.New, rawsecret)
-	case HmacSHA384:
-		h = hmac.New(sha512.New384, rawsecret)
-	case HmacSHA512:
-		h = hmac.New(sha512.New, rawsecret)
-	default:
-		return nil, "", ErrKeyAlg
-	}
-	h.Write(buf)
 	// Copy all TSIG fields except MAC and its size, which are filled using the computed digest.
 	*t = *rr
-	t.MAC = hex.EncodeToString(h.Sum(nil))
+	if provider == nil {
```

Comment:
Instead of having to special case `provider == nil` what if we provide a really simple (unexported) HMAC provider and use that throughout?

Basically this:
```go
type tsigHMACProvider string // The secret key to verify with.

func (key tsigHMACProvider) Generate(...) ([]byte, error) {
	secret, err := fromBase64([]byte(key))
	// HMAC stuff here
}

func (key tsigHMACProvider) Verify(...) { ... }

...
```

Then `TsigGenerate` and `TsigVerify` just create it unconditionally and `ReadMsg` and `WriteMsg` use it if `TsigProvider` is `nil`. I think that should make this much cleaner.

Reply:
I think that's done now.